### PR TITLE
Add an iam role session duration parameter to AWS module

### LINKED ISSUE #9086: Add the possibility to define the session duration when assuming roles on AWS-S3 wodle
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| Any | AWS-S3 Wodle | Manager/Agent | Packages/Sources | - |

Hello team,

This issue aims to add the possibility to define the session duration when assuming a role on the AWS-S3 wodle.

As stated in [boto3 documentation - sts.client.assume_role](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts.html#STS.Client.assume_role):

> **DurationSeconds (integer) --**
> 
> The duration, in seconds, of the role session. The value specified can can range from 900 seconds (15 minutes) up to the maximum session duration that is set for the role. The maximum session duration setting can have a value from 1 hour to 12 hours. If you specify a value higher than this setting or the administrator setting (whichever is lower), the operation fails. For example, if you specify a session duration of 12 hours, but your administrator set the maximum session duration to 6 hours, your operation fails. To learn how to view the maximum value for your role, see View the Maximum Session Duration Setting for a Role in the IAM User Guide .

### Modifications

**Current implementation:**
```python
sts_role_assumption = sts_client.assume_role(RoleArn=iam_role_arn,
                                             RoleSessionName='WazuhLogParsing'
                                             )
```

**Possible iteration:**
```python
sts_role_assumption = sts_client.assume_role(RoleArn=iam_role_arn,
                                             RoleSessionName='WazuhLogParsing',
                                             DurationSeconds=iam_role_duration
                                             )
```

Regards,
Franco Hielpos

### PR DESCRIPTION
**Pull request by noise-kngdm**
|Related issue|
|---|
|Closes #9086 |
## Description
In this PR we have added a new parameter to the AWS integration module that allows users to set the duration of the session assumed when using the `--iam_role_arn` parameter. After the changes, the user can specify a session duration between 900 and 3600 seconds using either the `--duration-seconds` parameter or modifyng the `ossec.conf` file.

### Manual tests performed
To test the changes we temporary added some code to the module that would add a delay after the `assume_role` call to test if the module would return a timeout when trying to use a session after it has expired.
The execution did return an error, as expected.
### Testing a bucket
```
root@ff5389d2ebf2:/var/ossec/wodles/aws# /var/ossec/framework/python/bin/python3 ./aws-s3.py -b cloudtrail-bucket -t cloudtrail -ds 900 -i arn:aws:iam::xxxxxxx:role/test_role
Time now: 09:03:43
Role assumed. role_session_duration variable is set to 900. Sleeping for 910 seconds.
Time after sleep: 09:18:53
ERROR: Invalid credentials to access S3 Bucket
```
### Testing a service
```
root@ff5389d2ebf2:/var/ossec/wodles/aws# py ./aws-s3.py -sr inspector -ds 900 -i arn:aws:iam::xxxxxxx:role/test_role
Time now: 10:37:26
Role assumed. role_session_duration variable is set to 900. Sleeping for 910 seconds.
Time after sleep: 10:52:36
ERROR: An error occurred (ExpiredTokenException) when calling the ListFindings operation: The security token included in the request is expired
```

We have also tested that the new parameter works fine in the configuration of both a service and a bucket by adding the following configuration items (one at a time) to the `ossec.conf` file and modifying the python program to write the value of the `role_session_variable`, if specified, to a temporary file. 
#### Bucket testing configuration
```
<wodle name="aws-s3">
  <disabled>no</disabled>
  <interval>10m</interval>
  <run_on_start>yes</run_on_start>
  <skip_on_error>yes</skip_on_error>
  <bucket type="cloudtrail">
     <name>wazuh-test-bucket</name>
     <access_key>XXXXXX</access_key>
     <secret_key>XXXXXX</secret_key>
     <iam_role_arn>arn:aws:iam::XXXXX:role/test_role</iam_role_arn>
     <role_session_duration>1300</role_session_duration>
  </bucket>  
</wodle>
```
#### Service testing configuration
```
<wodle name="aws-s3">
  <disabled>no</disabled>
  <interval>1m</interval>
  <run_on_start>yes</run_on_start>
  <skip_on_error>yes</skip_on_error>
  <service type="cloudwatchlogs">
     <access_key>XXXXXX</access_key>
     <secret_key>XXXXXX</secret_key>
     <iam_role_arn>arn:aws:iam::XXXXX:role/test_role</iam_role_arn>
     <role_session_duration>1300</role_session_duration>
     <aws_log_groups>framework-test-apache</aws_log_groups>
  </service>  
</wodle>

### CODE REVIEW

File: wodles/aws/aws_s3.py [hunk 1]
```diff
@@ -3101,8 +3118,14 @@ def get_script_arguments():
                         help='URL for the VPC endpoint to use to obtain the STS token.')
     parser.add_argument('-se', '--service_endpoint', type=str, dest='service_endpoint', default=None,
                         help='URL for the VPC endpoint to use to obtain the logs.')
+    parser.add_argument('-ds', '--duration-seconds', type=int, dest='role_session_duration', default=None,
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3121]: Replace `-ds` and `--duration-seconds` with a more descriptive parameter name, such as `--iam_role_duration`.

File: wodles/aws/aws_s3.py [hunk 2]
```diff
@@ -3101,8 +3118,14 @@ def get_script_arguments():
...
+                        help='The duration, in seconds, of the role session. Value can range from 900s to the max'
+                        ' session duration set for the role.')
+    parsed_args = parser.parse_args()
 
-    return parser.parse_args()
+    if parsed_args.role_session_duration is not None and parsed_args.iam_role_arn is None:
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3126]: Create a new `arg_valid_iam_role_duration` function and handle this there instead

File: wodles/aws/aws_s3.py [hunk 3]
```diff
@@ -3112,6 +3135,9 @@ def get_script_arguments():
 def main(argv):
     # Parse arguments
     options = get_script_arguments()
+    if not tools.check_role_session_duration(options.role_session_duration):
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3138]: Perform this check in the suggested `arg_valid_iam_role_duration` function` function

File: wodles/aws/tools.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2015-2021, Wazuh Inc.
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
+
+def check_role_session_duration(duration: int or None) -> bool:
```
Comment (CarlosRS9) [inline wodles/aws/tools.py:6]: There is no need for a new `tools.py` module just for this function as this check can be performed in a `arg_valid_iam_role_duration` function in the `aws_s3.py` module

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

**Comment by CarlosRS9** [comment]
We should also try to use the same name for the ossec.conf tag option (`role_session_duration`) and the aws-s3.py parameter (`duration-seconds`) if possible

File: src/wazuh_modules/wm_aws.h [hunk 5]
```diff
@@ -28,6 +28,7 @@ typedef struct wm_aws_bucket {
     char *secret_key;                   // IAM secret key
     char *aws_profile;                  // AWS credentials profile
     char *iam_role_arn;                 // IAM role
+    char *iam_role_duration;        // IAM role session duration
```
Comment (CarlosRS9) [inline src/wazuh_modules/wm_aws.h:31]: Tabulate this accordingly

File: src/wazuh_modules/wm_aws.h [hunk 6]
```diff
@@ -51,6 +52,7 @@ typedef struct wm_aws_service {
...
```
Comment (CarlosRS9) [inline src/wazuh_modules/wm_aws.h:55]: Tabulate this accordingly

File: wodles/aws/aws_s3.py [hunk 7]
```diff
@@ -3044,6 +3058,23 @@ def arg_valid_regions(arg_string):
     return final_regions
 
 
+def arg_valid_iam_role_duration(duration: int or None, iam_role_arn: str or None):
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3061]:
```suggestion
def arg_valid_iam_role_duration(arg_string):
```

Argument validation functions will receive only one parameter when called

File: wodles/aws/aws_s3.py [hunk 8]
```diff
@@ -3044,6 +3058,23 @@ def arg_valid_regions(arg_string):
...
+    ----------
+    duration: int or None
+        The desired session duration in seconds.
+    iam_role_arn: str or None
+        The iam role arn provided.
+    """
+    # Session duration must be between 15m and 12h
+    if not (duration is None or (900 <= duration <= 3600)):
+        raise ValueError("Invalid session duration specified. Value must be between 900 and 3600.")
+    if duration is not None and iam_role_arn is None:
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3074]: Remove this. As you cannot access the `iam_role_arn` value from here you will need to move this condition to the `main` body. See the next comment for more details.

File: wodles/aws/aws_s3.py [hunk 9]
```diff
@@ -3044,6 +3058,23 @@ def arg_valid_regions(arg_string):
...
+    Parameters
+    ----------
+    duration: int or None
+        The desired session duration in seconds.
+    iam_role_arn: str or None
+        The iam role arn provided.
+    """
+    # Session duration must be between 15m and 12h
+    if not (duration is None or (900 <= duration <= 3600)):
+        raise ValueError("Invalid session duration specified. Value must be between 900 and 3600.")
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3073]: This should raise a `argparse.ArgumentTypeError` instead of `ValueError`

File: wodles/aws/aws_s3.py [hunk 10]
```diff
@@ -3113,6 +3149,7 @@ def main(argv):
     # Parse arguments
     options = get_script_arguments()
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3151]: Add the `iam_role_arn` validation from the `arg_valid_iam_role_duration` here:
```suggestion
if duration is not None and iam_role_arn is None:
    raise argparse.ArgumentTypeError('Used --iam_role_duration argument but no --iam_role_arn provided.')
```

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

File: wodles/aws/aws_s3.py [hunk 11]
```diff
@@ -3101,8 +3140,15 @@ def get_script_arguments():
                         help='URL for the VPC endpoint to use to obtain the STS token.')
     parser.add_argument('-se', '--service_endpoint', type=str, dest='service_endpoint', default=None,
                         help='URL for the VPC endpoint to use to obtain the logs.')
+    parser.add_argument('-rd', '--iam_role_duration', type=arg_valid_iam_role_duration, dest='iam_role_duration', default=None,
+                        help='The duration, in seconds, of the role session. Value can range from 900s to the max'
+                        ' session duration set for the role.')
+    parsed_args = parser.parse_args()
+
+    if parsed_args.iam_role_duration is not None and parsed_args.iam_role_arn is None:
+        raise Exception('Used --iam_role_duration argument but no --iam_role_arn provided.')
```
Comment (CarlosRS9) [inline wodles/aws/aws_s3.py:3149]:
```suggestion
        raise argparse.ArgumentTypeError('Used --iam_role_duration argument but no --iam_role_arn provided.')
```

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

**Review by CarlosRS9 — APPROVED** [review]

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**